### ISSUE TITLE
VNC Connection Capability

### ISSUE THREAD

**Issue by Riprock** [gravitational/teleport#13179]
### What would you like Teleport to do?
VNC connection type in the browser

### What problem does this solve?
Having desktop sessions on Linux machines via VNC.

### If a workaround exists, please include it.
No workaround in the browser to get VNC
Only work around is at the CLI and SSH port forwarding

### Use Case/reason for request
I help run competitions and the current method we are using for teams to access their infrastructure during the competition isnt working out well for us. So looking into different solutions and came across teleport. Only thing that it doesnt really do that meets our needs is have VNC capability hence this feature request.

**Comment by zmb3** [comment]
See also #12023

**Comment by ashleyepirus** [comment]
Is there an ETA for this feature?

**Comment by ibeckermayer** [comment]
@ashleyepirus @marcosimioni 

No news at this time but I’m hoping to get into it after the big performance improvement push that’s currently underway. In the meantime you can consider using VNC over Teleport App Access

**Comment by marcosimioni** [comment]
> @ashleyepirus @marcosimioni 
> 
> 
> 
> No news at this time but I’m hoping to get into it after the big performance improvement push that’s currently underway. In the meantime you can consider using VNC over Teleport App Access

Thanks for your update!

Can you please elaborate? What do you mean with VNC over Teleport App Access? Is there an HTTPS tunneled version of VNC you are referring to, that we can proxy via teleport as a regular web app?

**Comment by zmb3** [comment]
You can use [TCP App Access](https://goteleport.com/docs/application-access/guides/tcp/#step-34-start-app-proxy) with an app that looks something like this:

```
 - name: vnc
    description: "Connect to VNC server"
    uri: "tcp://server.example.com:5900"
    insecure_skip_verify: true
    labels:
      foo: bar
```

Then you'd use `tsh proxy app vnc` and point a VNC viewer at localhost.

**Comment by shizayn** [comment]
> @ibeckermayer
> 
> No news at this time but I’m hoping to get into it after the big performance improvement push that’s currently underway. In the meantime you can consider using VNC over Teleport App Access

Really keen for having xRDP supported as well!

**Comment by ibeckermayer** [comment]
@shizayn noted, thanks for the input. We haven't settled on VNC yet, adding support for xRDP would likely be easier since we already have an RDP stack integrated in Teleport.

**Comment by johansenha** [comment]
voting this feature request as well. Would love to get the integration without the VNC workaround.

**Comment by ibeckermayer** [comment]
@johansenha "without the VNC workaround" meaning you want the underlying protocol to be RDP rather than VNC?

**Comment by johansenha** [comment]
Native RDP support to connect to a Linux based distro running xrdp would be my favorite (including screen recording feature at best)

**Comment by johansenha** [comment]
Hello Team, any expectations for waiting customers on getting this feature integrated? :)

**Comment by acfranken** [comment]
Looking forward to it as well. Do you have any update?

**Comment by fuzzysteve** [comment]
Any information on how that experimentation went?

**Comment by ibeckermayer** [comment]
It went pretty well, discounting XRDP specifically. 

XRDP holds a lot of legacy baggage that make it difficult to integrate with Teleport's existing auth system. Currently we use a virtualized smart card to authenticate, and while there's [ongoing discussion](https://github.com/neutrinolabs/xrdp/discussions/2625) of building this in to XRDP, realistically it's beyond our control to push that timeline.

Another option could be a more standard x.509 cert-based auth integration, however XRDP's existing auth system could not presently support this. Again, there is an effort [to update that piece of the system](https://github.com/neutrinolabs/xrdp/discussions/1961), but again it's very complex and out of our control.

All that said, if we leave XRDP out of it, we were able to develop an agent-based architecture (closer to how Teleport's SSH access typically works) which uses an existing authenticated reverse tunnel made by the agent (which is simply the present `windows_desktop_service`) to the proxy and then creates an x11-based RDP server. It would take some time to build this up to feature parity (i.e. clipboard sharing, directory sharing, etc) with our windows system, but we were able to get basic mouse + keyboard + GUI going within the few days we were working on it.

Can't make any promises about the timeline but I'm hoping we will make it the top priority for next quarter.

**Comment by andrewbks** [comment]
An x11-based RDP server option sounds worthwhile.

**Comment by waldemar-kindler** [comment]
Would an x11-based RDP server have the same performance limitations as x11?

Have you also considered something similar to [KasmVNC](https://github.com/kasmtech/KasmVNC) or x2go, which both provide much better performance and native web clients?

**Comment by acfranken** [comment]
Yes please.

**Comment by twf0** [comment]
Any updates on the timeline ?

**Comment by zmb3** [comment]
> Any updates on the timeline ?

It is on the roadmap - we're probably looking at the second half of the year.

**Comment by Compr0mzd** [comment]
We are at it no ?

**Comment by manuveli** [comment]
> We are at it no ?

as it's open since 2022, I gave up and switched to Apache Guacamole. Perhaps that fits your needs, too

**Comment by solvingj** [comment]
I've just opened the related issue/request for macOS: 
https://github.com/gravitational/teleport/issues/59275

It is highly unfortunate that this thread fell into the quagmire of competing Linux Window technologies and nobody seemed to mention the fact that VNC would have worked for both Linux and macOS.  

@zmb3 would teleport be open to a community contributed implementation of VNC?

**Comment by zmb3** [comment]
> @zmb3 would teleport be open to a community contributed implementation of VNC?

No. This would be too big a feature to accept from the community, and we aren't confident that VNC will perform well enough in Teleport's architecture.

The team is prototyping several different implementations now. We're looking primarily at RDP to Linux hosts but also considering a custom solution.

**Comment by solvingj** [comment]
So glad I asked.  So, is it fair to update the MacOS issue to state there are no plans for native macOS desktop support at this time?

**Comment by twf0** [comment]
Any updates please ?

**Comment by yjperez** [comment]
> Any updates please ?

Actively being developed. Targeting release within the next quarter.

**Comment by pschisa** [comment]
+1 for GUI based Ubuntu system access

**Comment by lordvandal** [comment]
+1 for Linux GUI support - any updates on the roadmap or timeline please? Thanks!

**Comment by zmb3** [comment]
https://goteleport.com/docs/upcoming-releases/#linux-desktop-access

**Comment by zmb3** [comment]
Initial support for Linux is released: https://goteleport.com/docs/enroll-resources/desktop-access/linux-desktop/

We will expand the feature set over time, but we no longer need this issue for tracking.
